### DISCUSSION TITLE
Is it possible to force a single space for indentation?

### DISCUSSION THREAD

**Discussion by Hubro** [nvim-treesitter/nvim-treesitter#2740] [Q&A] [answered]
I'm working on implementing tree-sitter indentation for YANG files. I seem to have gotten 99% of the way with this:

```
(block) @indent
(block "}" @branch)

(statement) @indent
(statement ";" @indent_end)
```

This handles blocks beautifully, and it handles multi-line statements perfectly.

However, I am attempting to follow the code style used in IETF YANG models, see for example: https://github.com/YangModels/yang/blob/main/standard/ietf/RFC/iana-routing-types@2017-12-04.yang

Multi-line strings should start on the next line, indented once. However, successive lines should be indented by 1 additional space, as to line up with the starting quote. For example:

![image](https://user-images.githubusercontent.com/597206/161646113-94346bbf-ee64-4f50-bf0c-a177d40234cc.png)

See the IETF model above for many more examples.

- Is there any way to specify that indentation should be increased by 1 space?
- Considering that the entire string is a single token, is it possible to make a rule such as "indent the first line in a string once, indent the following lines one additional time"?

**Comment by theHamsta** [comment, marked as answer]
`@aligned` should work. Have a look on how it used in other indent queries.

**Reply by Hubro** [reply]
I'm unable to find anything similar done in any other indent queries. I thought perhaps I could find an example of somebody doing this kind of single space indentation:

```java
/*
 * I am indented by one space to align with the star
 * on the previous line
 */
```

But I couldn't find that either, not even for Java or JavaScript where this is super common. Most indent queries just use `@auto` or `@ignore` for comments :thinking:

**Reply by theHamsta** [reply]
Those are all one space indents with respect to "(" https://github.com/nvim-treesitter/nvim-treesitter/blob/56baf1873a42d54d018e92f1a0b493a51af7a131/queries/python/indents.scm#L43-L34

**Reply by theHamsta** [reply]
`"` can be escaped like that `"\"\""`

**Reply by Hubro** [reply]
Aah ok, I see. This seems to work:

```scm
((string) @aligned_indent
 (#set! "delimiter" "\"\""))
```

Thanks! :grin:

**Reply by Hubro** [reply]
@theHamsta Sorry for pestering you, but I noticed another special indentation case in the IETF models. Strings that are split into multiple lines and concatenated are aligned by their starting quote, for example:

```yang
leaf language {
  type string {
    pattern '(([A-Za-z]{2,3}(-[A-Za-z]{3}(-[A-Za-z]{3})'
          + '{,2})?|[A-Za-z]{4}|[A-Za-z]{5,8})(-[A-Za-z]{4})?'
          + '(-([A-Za-z]{2}|[0-9]{3}))?(-([A-Za-z0-9]{5,8}'
          + '|([0-9][A-Za-z0-9]{3})))*(-[0-9A-WY-Za-wy-z]'
          + '(-([A-Za-z0-9]{2,8}))+)*(-[Xx](-([A-Za-z0-9]'
          + '{1,8}))+)?|[Xx](-([A-Za-z0-9]{1,8}))+|'
          + '(([Ee][Nn]-[Gg][Bb]-[Oo][Ee][Dd]|[Ii]-'
          + '[Aa][Mm][Ii]|[Ii]-[Bb][Nn][Nn]|[Ii]-'
          + '[Dd][Ee][Ff][Aa][Uu][Ll][Tt]|[Ii]-'
          + '[Ee][Nn][Oo][Cc][Hh][Ii][Aa][Nn]'
          + '|[Ii]-[Hh][Aa][Kk]|'
          + '[Ii]-[Kk][Ll][Ii][Nn][Gg][Oo][Nn]|'
          + '[Ii]-[Ll][Uu][Xx]|[Ii]-[Mm][Ii][Nn][Gg][Oo]|'
          + '[Ii]-[Nn][Aa][Vv][Aa][Jj][Oo]|[Ii]-[Pp][Ww][Nn]|'
          + '[Ii]-[Tt][Aa][Oo]|[Ii]-[Tt][Aa][Yy]|'
          + '[Ii]-[Tt][Ss][Uu]|[Ss][Gg][Nn]-[Bb][Ee]-[Ff][Rr]|'
          + '[Ss][Gg][Nn]-[Bb][Ee]-[Nn][Ll]|[Ss][Gg][Nn]-'
          + '[Cc][Hh]-[Dd][Ee])|([Aa][Rr][Tt]-'
          + '[Ll][Oo][Jj][Bb][Aa][Nn]|[Cc][Ee][Ll]-'
          + '[Gg][Aa][Uu][Ll][Ii][Ss][Hh]|'
          + '[Nn][Oo]-[Bb][Oo][Kk]|[Nn][Oo]-'
          + '[Nn][Yy][Nn]|[Zz][Hh]-[Gg][Uu][Oo][Yy][Uu]|'
          + '[Zz][Hh]-[Hh][Aa][Kk][Kk][Aa]|[Zz][Hh]-'
          + '[Mm][Ii][Nn]|[Zz][Hh]-[Mm][Ii][Nn]-'
          + '[Nn][Aa][Nn]|[Zz][Hh]-[Xx][Ii][Aa][Nn][Gg])))';
  }
  default "en-US";
}
```

There isn't really a starting delimiter to align by here, like there is with parameter lists and multi-line strings.

Is there any way to express this type of indentation with indent queries?

---

The syntax tree looks like this:

![image](https://user-images.githubusercontent.com/597206/161732106-8efaf8e0-cce0-4f29-9772-b796720da8fe.png)

**Reply by theHamsta** [reply]
The `@aligned_indent` has a secret parameter `increment` https://github.com/nvim-treesitter/nvim-treesitter/blob/56baf1873a42d54d018e92f1a0b493a51af7a131/lua/nvim-treesitter/indent.lua#L180 so you could experiment with `(#set increment -3)` if tree-sitter can parse negative values (not sure about that, probably also needs a check that you can't get negative with indent).

**Reply by Hubro** [reply]
Aha, that's cool :thinking: But the main issue I see is that I don't have a clear start/end delimiter for the aligned indent, so I'm not sure how I would even begin.
